refactor(metrics): export pipeline

legendecas opened the pull request:
Metrics exporting requires a clear policy about when to export.  This PR extracts controllers from `metrics.Meter` and makes it installable on `metrics.MeterProvider`.

## Which problem is this PR solving?

- MetricsProvider should act as a Controller to determine when to collect all metrics values.
- Related #796.
- Go impl: https://github.com/open-telemetry/opentelemetry-go/blob/a485d0ec64a48f6b7d5344790ba6c4b85f154e8e/sdk/metric/doc.go#L156

## Short description of the changes

- `metrics.MeterProvider` exposes a new method `collect` that similar to `metrics.Meter.collect`.
- Update the PushController to accept a `metrics.MeterProvider` but not a `metrics.Meter`.
- Add public `installExportPipeline` function to @opentelemetry/exporter-prometheus.
- Add public `installExportPipeline` function to ConsoleExporter in @opentelemetry/metrics to install a push controller on global `api.metrics`.

vmarchaud commented on packages/opentelemetry-metrics/src/export/controllers/PushController.ts:
```diff
@@ -0,0 +1,73 @@
+import { UngroupedBatcher } from '../Batcher';
+import { Controller } from '../Controller';
+
+const DEFAULT_EXPORT_INTERVAL = 60_000;
```
think this should be `60000` ?

legendecas replied:
Yeah, quite the same. This syntax is a [stage 3 proposal](https://github.com/tc39/proposal-numeric-separator) that already working in transpilers like TypeScript.

legendecas replied:
This is not new code, previously at https://github.com/open-telemetry/opentelemetry-js/blob/master/packages/opentelemetry-metrics/src/export/Controller.ts#L21 moved here.

vmarchaud replied:
Didn't knew about this syntax, thanks for explaining it :)

mayurkale22 commented:
I am little confused about the this PR (esp. `MeterProvider` -> `PushController`), is this something enforced by the specs?

Currently, `Meter` instance encapsulates the Controller logic. I think it should only be responsible for creating and recording metrics. And we should have separate export pipeline using either `PushController` or `PullController`.

Basically, users have to do the following things:
1. Create meter using MeterProvider
   `const meter = new MeterProvider();`
2. Define the Contoller which will take meter and exporter to export the data at some interval.
   `const controller = new PushController(meter, exporter, interval);`
   OR 
	`const controller = new PullController(meter, exporter);`

mayurkale22 commented:
Looks like this is similar to Python's implementation. See: https://github.com/open-telemetry/opentelemetry-python/blob/master/docs/examples/basic_meter/basic_metrics.py#L66

legendecas commented:
@mayurkale22  an alternative could be the pattern opentelemetry-go is using:

(the example is written in JavaScript to illustrate possible API shapes)
```js
import prometheus from '@opentelemetry/exporter-prometheus'
import { metrics } from '@opentelemetry/api'

prometheus.installPipeline(); // registers the PullController as globalMeterProvider.
const meter = metrics.getMeter('example-meter');
const counter = meter.createCounter('foo');
// ...
```

In this example, end users of `@opentelemetry/api`/`@opentelemetry/metrics` may not aware the controller/MeterProvider, these are taken care of by the exporters since those exporters do care about how the metrics are going to be collected.

The python example does work. But dangling a variable like `pushController` doesn't seem to be a best practice to me (they are ref-ed by hosts, cool, but confusing).

WDYT? @mayurkale22 @obecny

dyladan commented:
so the exporter decides to install push or pull controller as the global meter provider?

vmarchaud commented:
Woudn't it mean that one meter provider can only be pull or push based, so you wouldn't be able to use two exporters at the same time that use each method ?

legendecas commented:
> so the exporter decides to install push or pull controller as the global meter provider?

Yes, push or pull strategy is based on how the metrics are exported. So I think this is more intuitive.

> Woudn't it mean that one meter provider can only be pull or push based, so you wouldn't be able to use two exporters at the same time that use each method ?

As a matter of fact, MeterProvider in @opentelemetry/metrics doesn't support multiple exporters already. Though the exporter can act like MultiSpanProcessor passing the metric records to all subsidiaries.

legendecas commented:
@dyladan @mayurkale22  @obecny @vmarchaud  hey, I've updated the PR to address the comments. Maybe there can be another round of review?

vmarchaud commented on packages/opentelemetry-metrics/src/export/controllers/Controller.ts:
```diff
@@ -0,0 +1,44 @@
+
+  protected onExportResult(result: ExportResult): void {
+    if (result !== ExportResult.SUCCESS) {
+      // @todo: log error
```
I think it's worth to add a logger option while are it, maybe as a optional parameter for now, WYTD ?

dyladan commented on packages/opentelemetry-metrics/src/export/controllers/Controller.ts:
```diff
@@ -0,0 +1,46 @@
+import { MetricExporter } from '../types';
+import { Batcher } from '../Batcher';
+
+export abstract class Controller {
```
Why is this abstract? It doesn't have any abstract properties

dyladan commented on packages/opentelemetry-exporter-prometheus/src/prometheus.ts:
```diff
@@ -117,6 +118,15 @@ export class PrometheusExporter implements MetricExporter {
+   *
+   * @param cb The callback
+   */
+  setPullCallback(cb?: () => void) {
```
We should specify in the tsdoc that it should be synchronous, otherwise the metrics will be exported while the pull callback is still running.

dyladan commented on packages/opentelemetry-metrics/src/export/controllers/PushController.ts:
```diff
@@ -0,0 +1,77 @@
+   *
+   * @returns Meter A Meter with the given name and version
+   */
+  getMeter(name: string, version = '*', config?: MeterConfig): Meter {
```
Why is `getMeter` implemented twice instead of being included in the base class?

legendecas commented:
> @dyladan: Minor issue: with the naming change, it may not be obvious to users that the controller is the meter provider.

I've appended more ts-doc on controllers to explicitly indicate that those controllers are capable of been installed as the global metric provider.

obecny commented on packages/opentelemetry-metrics/README.md:
```diff
@@ -11,11 +11,26 @@ OpenTelemetry metrics allow a user to collect data and export it to a metrics ba
+import { metrics } from '@opentelemetry/api'
+
+// Install the export pipeline before all subsequent call to metrics.
+ConsoleMetricExporter.installPipeline();
```
why is this static, have different name than in prometheus and the flow will also looks differently ?
I think this all should look the same doesn't matter if this is console exporter, prometheus exporter or any other exporter. I think we should have then one method in metrics which will install the `export pipeline` by accepting an exporter as dependency injection.

legendecas replied:
I'll rename those `installExportPipeline` aligned if the above explanation make sense to you.

dyladan commented:
Maybe we can resolve naming confusion by naming the controllers `PullControllerMeterProvider` and `PushControllerMeterProvider`.

obecny commented:
> Maybe we can resolve naming confusion by naming the controllers `PullControllerMeterProvider` and `PushControllerMeterProvider`.

I would keep the naming the same as in spec to avoid any confusion (MeterProvider) but then maybe some option can decide if that is a push or pull  strategy ?. 
```
const provider = new MeterProvider('pull'); // default to 'push' as it was
const provider = new MeterProvider({controller: 'pull'}); // default to 'push' as it was
const provider = new MeterProvider({strategy: 'pull'}); // default to 'push' as it was
...

legendecas commented on packages/opentelemetry-exporter-prometheus/src/export/export-pipeline.ts:
```diff
@@ -0,0 +1,38 @@
+ * limitations under the License.
+ */
+
+import * as api from '@opentelemetry/api';
```
The key point for `installExportPipeline` is that the export policy mainly depends on how the exporter works. For example, the prometheus exporter (which starts a server awaiting incoming scrapping) would prefer a pulling strategy and the console exporter would prefer a periodically print of the metric records. For this reason, the `installExportPipeline` was implemented for each exporter and in the `installExportPipeline` there is nothing magic other than installing a metric exporting strategy (controller) as the global metric provider as the metric provider is the role who holds all meters/metrics.

If the `installExportPipeline` is implemented as a method in opentelemetry/metric or core, there is no significant difference with the current approach since the exporting strategy still needs to be determined by who installs the pipeline which can be probably the end-user who uses the exporter rather than the exporter themselves.

e.g.

```js
metrics.installExportPipeline({ exporter: new PrometheusExporter({ startServer: true }), strategy: 'pull' });
```

The assumption is based on that the end-user doesn't have to care about the strategy the exporter used. They do care about which exporter they are going to work with.

dyladan commented:
What is the status of this? It has 3 reviews, but open comments from a maintainer.

@obecny you are the only one with remaining open comments here.

obecny commented:
> What is the status of this? It has 3 reviews, but open comments from a maintainer.
> 
> @obecny you are the only one with remaining open comments here.

My main concern is that why this is done different then what spec says. Spec says about MeterProvider, not PushController.
I think @mayurkale22 had the same concerns.

legendecas commented:
@obecny I've updated the PR the revert the changes made to MeterProvider naming. 

Since this update, the PR can be summarized as:

- `metrics.MeterProvider` exposes a new method `collect` that similar to `metrics.Meter.collect`, collecting all metrics of all meters that the meter provider creates.
- Renamed Controller to PushController, and the new PushController accepts a `metrics.MeterProvider` but not a `metrics.Meter`.
- Add public `installExportPipeline` function to `@opentelemetry/exporter-prometheus`.
- Add public `installExportPipeline` function to `ConsoleExporter` in `@opentelemetry/metrics` to install a push controller on global `api.metrics`.

Example usage of the new pipeline:

```js

import { ConsoleMetricExporter } from '@opentelemetry/metrics'
import { metrics } from '@opentelemetry/api'

// Install the export pipeline before all subsequent call to metrics.
ConsoleMetricExporter.installExportPipeline();
// Identical to:
(function installExportPipeline() {
  const exporter = new ConsoleMetricExporter();
  const meterProvider = new MeterProvider({ ...meterConfig, exporter });
  const controller = new PushController(meterProvider, config);
  api.metrics.setGlobalMeterProvider(meterProvider);
})();

const meter = metrics.getMeter('example-meter');
const counter = meter.createCounter('foo');
```

obecny commented on packages/opentelemetry-metrics/src/MeterProvider.ts:
```diff
@@ -14,27 +14,53 @@
+            result
+          );
+        }
+        resolve();
```
Curious why only resolve without reject ?

legendecas replied:
Moved the export result handling to the push controller so that we can implement retry with controllers.

obecny commented on packages/opentelemetry-metrics/src/export/ConsoleMetricExporter.ts:
```diff
@@ -15,13 +15,32 @@
+  static installExportPipeline(
+    config?: PushControllerConfig,
+    meterConfig?: MeterConfig
+  ) {
```
missing return type - I think it should be the same interface that you have in prometheus, we might use it later for CollectorMetricExporter as well

obecny commented on packages/opentelemetry-metrics/test/MeterProvider.test.ts:
```diff
@@ -21,55 +21,55 @@ import { NoopLogger } from '@opentelemetry/core';
     it('should construct an instance without any options', () => {
-      const provider = new MeterProvider();
-      assert.ok(provider instanceof MeterProvider);
+      const controller = new MeterProvider();
```
I think you can rename it back to provider in all tests ?

legendecas commented on packages/opentelemetry-exporter-prometheus/src/prometheus.ts:
```diff
@@ -42,13 +43,32 @@ export class PrometheusExporter implements MetricExporter {
+    exporterConfig?: ExporterConfig,
+    callback?: () => void,
+    controllerConfig?: ConstructorParameters<typeof MeterProvider>[0]
+  ) {
```
Updated. Added a new interface `ExportPipelineInstaller` to describe the `installExportPipeline` method.

dyladan commented:
First, thank you for the work you have put into this. I know this has been a lot and the review process has been lengthy, but this is a large PR and there is a lot to unpack here. First, I have a couple of questions.

1. Why did you decide to go with this model, rather than the simpler route of adding a pull controller to the existing model? This is a large change from what we previously had. A smaller change would be something like this:

```typescript
interface PullingExporter extends Exporter {
  registerPullController(): void;
}


const pullingExporter = new Exporter1(); // implements pulling exporter
const pushingExporter = new Exporter2(); // implement exporter

const provider = new MeterProvider();

provider.addController(new PullController(pullingExporter));
provider.addController(new PushController(pushingExporter);

provider.register();
```

In addition to being simpler, this would also have good symmetry with the tracing API.

2. This looks similar to the go and python implementations with one notable exception. As you can see [here](https://github.com/open-telemetry/opentelemetry-python/blob/master/docs/examples/basic_meter/basic_metrics.py#L51), the export pipeline installation method is on the `MeterProvider` in Python, where you decided to put it in the exporter. Can you please explain why you decided to add the function to the exporter rather than the provider? I can easily see a situation where the user writes the following code and is confused why it doesn't work:

```typescript
const provider = new MeterProvider();
metrics.setMeterProvider(provider);

const meter = provider.getMeter("myapp");

const exporter = new Exporter();
exporter.installExportPipeline();

const counter = meter.createCounter();

// use counter in some way, metrics are not exported, user is confused
```

Because this PR has the `installExportPipeline` create a _new_ `MeterProvider`, the user's counter in this case would not work with the export pipeline and the user would see no metrics. In some ways, this is a documentation issue, but we should also do our best to avoid situations like this where it is so easy to misuse the API.

legendecas commented:
@dyladan Thanks for the questions.

The idea behind the `metricExporter.installExportPipeline` is that metric exporter has to decide their metric collection frequency to reduce costs (reduce proactive metric collections like system resources utilization) at their best efforts. My experience is some APM vendors do only collect metrics with a limit set of time periods, like 5s, or 15s. If the pipeline is installed by users, like the python example, the users have to decide the intervals of metric collections. In that way, the exporter may have to implement their own caching mechanism to not flushing their servers if the user sets quite a short interval. However, if the exports do support arbitrary intervals (or strategies), they can provide options to alter that if the `installExportPipeline` is implemented by the exporters.

However, I found I'm fallen into a thought trap while reviewing the first question. My initial thought is that we can lift the decision like the strategy the exporter should use. Yet I found the symmetry between the metrics and tracing you mentioned is interesting idea on how the metrics export pipeline should be defined, regardless of how currently go or python SDK is implemented.

My thoughts on the symmetry can be: Aggregator & Controller <-> Sampler. Which aggregator to use is decided by the batcher, which is installed at the construction of MeterProvider. Similarly, samplers will be installed on the construction of BasicTracerProvider. So for controllers, I'm wondering if the current approach is the right way to keep symmetry between metrics and tracing. I'm opening to revert the current changes to only migrating Controllers on the main branch to bind with `MeterProvider` rather than single `Meter` instance (so that this PR can be split into smaller ones), and continue the export pipeline installation discussion in another PR (or issue).

legendecas commented:
The export pipeline user interface may be hard to be a stable design as the metric record export pipeline of metrics SDK specification is still been drafting actively. Closed now and keep tracking on https://github.com/open-telemetry/opentelemetry-js/issues/1528.
